pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go

```diff
@@ -0,0 +1,498 @@
+			return nil
+		}
+		return fmt.Errorf("Error retrieving CloudPrivateIPConfig: %s from the API server, err: %v", name, err)
+	}
+
+	nodeToAdd, nodeToDel := c.computeOp(cloudPrivateIPConfig)
+
+	// Dequeue on NOOP, there's nothing to do
+	if nodeToAdd == "" && nodeToDel == "" {
+		return nil
+	}
+
+	if nodeToDel != "" {
+
+		klog.Infof("CloudPrivateIPConfig: %s will be deleted from node: %s", name, nodeToDel)
+		ip := net.ParseIP(cloudPrivateIPConfig.Name)
+
+		node, err := c.nodesLister.Get(nodeToDel)
+		if err != nil {
+			return fmt.Errorf("corev1.Node: %s could not be retrieved from the API server, err: %v", node.Name, err)
+		}
+
+		if cloudRequestObj, err = c.CloudProviderClient.ReleasePrivateIP(ip, node); err != nil {
+			// Delete operation encountered an error, requeue
+			status = &cloudnetworkv1.CloudPrivateIPConfigStatus{
+				Node: cloudPrivateIPConfig.Status.Node,
+				Conditions: []metav1.Condition{
+					metav1.Condition{
+						Type:               string(cloudnetworkv1.Assigned),
+						Status:             metav1.ConditionFalse,
+						ObservedGeneration: cloudPrivateIPConfig.Status.Conditions[0].ObservedGeneration + 1,
+						LastTransitionTime: metav1.Now(),
+						Reason:             cloudResponseReasonError,
+						Message:            fmt.Sprintf("Error issuing cloud release request, err: %v", err),
+					},
+				},
+			}
+			// Always requeue the object if we end up here. We need to make sure
+			// we try to clean up the IP on the cloud
+			if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				cloudPrivateIPConfig, err = c.updateCloudPrivateIPConfigStatus(cloudPrivateIPConfig, status)
+				return err
+			}); err != nil {
+				return fmt.Errorf("Error updating CloudPrivateIPConfig: %s during delete operation, err: %v", name, err)
+			}
+			return fmt.Errorf("CloudPrivateIPConfig: %s could not be released from node: %s, err: %v", name, node.Name, err)
+		}
+		// This is step 2. in the docbloc for the DELETE operation in the
+		// syncHandler
+		status = &cloudnetworkv1.CloudPrivateIPConfigStatus{
+			Node: cloudPrivateIPConfig.Status.Node,
+			Conditions: []metav1.Condition{
+				metav1.Condition{
+					Type:               string(cloudnetworkv1.Assigned),
+					Status:             metav1.ConditionUnknown,
+					ObservedGeneration: cloudPrivateIPConfig.Status.Conditions[0].ObservedGeneration + 1,
+					LastTransitionTime: metav1.Now(),
+					Reason:             cloudResponseReasonPending,
+				},
+			},
+		}
+		if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			cloudPrivateIPConfig, err = c.updateCloudPrivateIPConfigStatus(cloudPrivateIPConfig, status)
+			return err
+		}); err != nil {
+			return fmt.Errorf("Error updating CloudPrivateIPConfig: %s during delete operation, err: %v", name, err)
+		}
+		// This is a long running and blocking function call.
+		cloudErr := c.CloudProviderClient.WaitForResponse(cloudRequestObj)
```

Comment:
Is it OK that we're blocked on other requests while this is running? Does it make sense to try and make this code thread-safe, and thus allow for multiple in-flight requests?

(I noticed that the AWS implementation does some checking that IPs are in a desired state; that might not work if running in parallel)

Reply:
This code is thread-safe and does allow for multiple in-flight requests. `defaultWorkerThreadiness` is 10, which means we can process 10 simultaneous items off of the workqueue.